packages/flutter/lib/src/widgets/view.dart

```diff
@@ -106,13 +145,669 @@ class View extends StatelessWidget {
+
+  void _attachToViewHooks([ViewHooks? viewHooks]) {
+    assert(_attachmentPoint == null);
+    viewHooks ??= ViewHooks.of(this);
+    viewHooks.pipelineOwner.adoptChild(_effectivePipelineOwner);
+    viewHooks.renderViewManager.addRenderView(renderObject);
+    _attachmentPoint = viewHooks;
+  }
+
+  void _detachFromViewHooks() {
+    final ViewHooks? viewHooks = _attachmentPoint;
+    if (viewHooks != null) {
+      viewHooks.renderViewManager.removeRenderView(renderObject);
+      viewHooks.pipelineOwner.dropChild(_effectivePipelineOwner);
+      _attachmentPoint = null;
+    }
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    if (_attachmentPoint == null) {
+      return;
+    }
+    final ViewHooks newHooks = ViewHooks.of(this);
+    if (newHooks != _attachmentPoint) {
+      _detachFromViewHooks();
+      _attachToViewHooks(newHooks);
+    }
+  }
+
+  @override
+  void performRebuild() {
+    super.performRebuild();
+    _updateChild();
+  }
+
+  @override
+  void activate() {
+    super.activate();
+    assert(_effectivePipelineOwner.rootNode == null);
+    _effectivePipelineOwner.rootNode = renderObject;
+    _attachToViewHooks();
+  }
+
+  @override
+  void deactivate() {
+    _detachFromViewHooks();
+    assert(_effectivePipelineOwner.rootNode == renderObject);
+    _effectivePipelineOwner.rootNode = null; // To satisfy the assert in the super class.
+    super.deactivate();
+  }
+
+  @override
+  void update(RawView newWidget) {
+    super.update(newWidget);
+    _updateChild();
+  }
+
+  @override
+  void visitChildren(ElementVisitor visitor) {
+    if (_child != null) {
+      visitor(_child!);
+    }
+  }
+
+  @override
+  void forgetChild(Element child) {
+    assert(child == _child);
+    _child = null;
+    super.forgetChild(child);
+  }
+
+  @override
+  void insertRenderObjectChild(RenderBox child, Object? slot) {
+    assert(slot == null);
+    assert(renderObject.debugValidateChild(child));
+    renderObject.child = child;
+  }
+
+  @override
+  void moveRenderObjectChild(RenderObject child, Object? oldSlot, Object? newSlot) {
+    assert(false);
+  }
+
+  @override
+  void removeRenderObjectChild(RenderObject child, Object? slot) {
+    assert(slot == null);
+    assert(renderObject.child == child);
+    renderObject.child = null;
+  }
+
+  @override
+  void unmount() {
+    if (_effectivePipelineOwner != (widget as RawView)._deprecatedPipelineOwner) {
+      _effectivePipelineOwner.dispose();
+    }
+    super.unmount();
+  }
 }
 
 class _ViewScope extends InheritedWidget {
   const _ViewScope({required this.view, required super.child});
 
-  final FlutterView view;
+  final FlutterView? view;
 
   @override
   bool updateShouldNotify(_ViewScope oldWidget) => view != oldWidget.view;
 }
+
+/// Injects [ViewHooks] into the widget tree so that they can be looked up by
+/// descendants via [ViewHooks.of].
+class ViewHooksScope extends InheritedWidget {
```

Comment:
Does this need to be public?

Reply:
From my understanding, this allows anyone to create widgets that create their own render trees, but it's not clear to me either which customer scenarios might need this.

Reply:
Huh, interesting find. This indeed can (and probably should) be private. Currently, the only usage preventing this from going private is that I use it in a couple of tests. I am taking an AI to restructure those tests so we can make this private. Thanks for pointing this out, I totally missed it!

> From my understanding, this allows anyone to create widgets that create their own render trees, but it's not clear to me either which customer scenarios might need this.

Also not sure what the use case for this would be. If we find one where it makes sense we can always make this public again. Although I would suspect that if you define your own render-tree creating widget you'd also want to replace the ViewHooks stuff with whatever makes sense for your new widget. There shouldn't really be a need to re-use these ViewHooks...